arch/xtensa/src/esp32/esp32_rmt.c

```diff
@@ -0,0 +1,315 @@
+  
+      esp32_teardown_irq(dev->cpu, dev->periph, dev->cpuint);
+      dev->cpuint = -ENOMEM;
+      leave_critical_section(flags);
+  
+      return ret;
+    }
+
+  /* Enable the CPU interrupt that is linked to the RMT device. */
+  up_enable_irq(dev->cpuint);
+
+  leave_critical_section(flags);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: esp32rmt_interrupt
+ *
+ * Description:
+ *   RMT TX interrupt handler
+ *
+ * Input Parameters:
+ *   irq - The IRQ number of the interrupt.
+ *   context - The register state save array at the time of the interrupt.
+ *   arg - The pointer to driver structure.
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno on failure
+ *
+ ****************************************************************************/
+IRAM_ATTR static int esp32rmt_interrupt(int irq, void *context, void *arg)
+{
+  struct rmt_dev_s *dev = (struct rmt_dev_s *)arg;
+  uint32_t regval = getreg32(RMT_INT_ST_REG);
+  uint32_t total_pixels;
+  uint32_t current_pixel;
+  uint32_t dst_mem;
+  uint32_t used_words;
+  uint8_t *pixel_ptr;
+  
+  if (regval & RMT_CHn_TX_END_INT_ST(dev->channel))
+    {
+      putreg32(RMT_CHn_TX_END_INT_CLR(dev->channel), RMT_INT_CLR_REG);
+      regval = getreg32(RMT_INT_ENA_REG);
+      regval &= ~(RMT_CHn_TX_END_INT_ENA(dev->channel));
+      putreg32(regval, RMT_INT_ENA_REG);
+      nxsem_post(&dev->tx_sem);
+    }
+  else if (regval & RMT_CHn_TX_THR_EVENT_INT_ST(dev->channel))
+    {
+      putreg32(RMT_CHn_TX_THR_EVENT_INT_CLR(dev->channel), RMT_INT_CLR_REG);
+      total_pixels = dev->data_len;
+      current_pixel = dev->next_pixel_to_load;
+      dst_mem = dev->start_address;
+      if(dev->frame++==1)
+        dst_mem += dev->reload_thresh * 4;
+      
+      if (dev->frame > 1)
+        dev->frame = 0;
+      
+      used_words = 0;
+      pixel_ptr = dev->data_ptr + current_pixel;
+  
+      while (used_words < dev->reload_thresh)
+        {
+          if (current_pixel < total_pixels)
+            {
+              register uint8_t pixel = (*pixel_ptr++);
+              for (register uint32_t i = 0; i < 8; i++)
+                {
+                  if (pixel & 0x80L)
+                    {
+                      putreg32(LOGIC_ONE, dst_mem);
+                    }
+                  else
+                    {
+                      putreg32(LOGIC_ZERO, dst_mem);
+                    }
+                  pixel <<= 1;
+                  dst_mem += 4;
+                }
+              used_words+=8;
+              current_pixel++;
+            }
+          else
+            {
+              regval = getreg32(RMT_INT_ENA_REG);
+              regval |= RMT_CHn_TX_END_INT_ENA(dev->channel);
+              putreg32(regval, RMT_INT_ENA_REG);
+              putreg32(0, dst_mem);
+              break;
+            }
+        }
+      
+      dev->next_pixel_to_load = current_pixel;
+    }
```

Comment:
I am not very familiar with the WS2812 nor the RMT peripheral.
Could you please give a brief explanation of what this block is implementing?

I have a feeling that this is application-specific logic and should not be present in driver code.

Reply:
If that is the case, this application-specific logic should be called via a callback, similarly to how an application is notified of a Timer expiration.
https://nuttx.apache.org/docs/latest/components/drivers/character/timer.html#c.TCIOC_NOTIFICATION

Reply:
The ws2812 is an 1-wire RGB LED that can be daisy chained with many others. Each LED receives a train of pulses encoding the 0's or 1's required to encode 24 bits of RGB color. Once the LED receives its pulses, it will adopt that color and only forward the remaining pulses to the next LED.
Pulses for zeros and ones have an specific timing each, so LOGIC_ZERO and LOGIC_ONE will set the RMT to generate the output wave that will encode the desired bit in the output GPIO.
Each channel of ESP32's RMT can hold up to 64 words of data, and we use one word to encode each pulse. This is enough for a single LED, however, when we have many LEDs connected to the system, there is the need to keep "refueling" the RMT's input with more data. 
The RMT will generate an interrupt once it has transmitted half of the data on its internal buffer, so, when this happens, this routine will copy more data to its buffer. This process will happen until all the input data is copied to RMT's internal buffer. 
In the last "refueling" interaction, 0 is written to RMT's buffer, and once RMT's transmission state machine reaches that point, it will go back to IDLE and generate an END interrupt, so that's when the write lock is released and the write function finally returns.

Perhaps LOGIC_ONE and LOGIC_ZERO definitions' can be considered application specific and could be members of the dev structure. Also the term pixel, doesn't help much. I've considered translating the output bytes to the RMT encoding format inside the write function and then we would have no need to translate the bytes to pulse bits inside the interruption, however, it would require a lot of extra memory, since a single bit of input data, becomes a word when transmitting.

Reply:
Thanks for the explanation!

> Perhaps LOGIC_ONE and LOGIC_ZERO definitions' can be considered application specific and could be members of the dev structure. Also the term pixel, doesn't help much. I've considered translating the output bytes to the RMT encoding format inside the write function and then we would have no need to translate the bytes to pulse bits inside the interruption, however, it would require a lot of extra memory, since a single bit of input data, becomes a word when transmitting.

The issue is that this whole logic is application-specific, in this case for blinking LEDs controlled by the WS2812. But RMT could be used for different purposes. IDF, for example, provides several examples besides LED blinking: https://github.com/espressif/esp-idf/tree/master/examples/peripherals/rmt

So this whole logic should not be present in the RMT code. Instead, it could be placed either at a device driver that uses the RMT or even at the application layer.
See how the Timer driver provides an interface for registering a callback for handling timer expiration:
https://github.com/apache/incubator-nuttx/blob/985710665a62e886694cf87f00be08fdf38e4702/arch/xtensa/src/esp32/esp32_tim_lowerhalf.c#L156-L179